feat(css): CSS is pre-processed by css-crush, all /cache output is filterable

mrclay opened the pull request:
All `/cache` output is now filtered by either the `simplecache:generate` or `cache:generate` hook, depending on whether Simplecache is enabled.

css-crush is used to pre-process all CSS `/cache` output.

Fixes #10625
Refs #10316

- [ ] test basic pre-processing

(replaces #10666)

mrclay commented:
FYI with simplecache off, elgg.css takes ~200ms longer to load (or 304). This is still 140ms faster than the initial PR, and seems acceptable for development.

jdalsem commented on docs/guides/upgrading.rst:
```diff
@@ -401,6 +401,8 @@ Notable changes in plugins:
  * Admin theme now reuses icon classes from ``elements/icons.css``
 
+Also note, CSS views served via ``/cache`` URLs are pre-processed using `CSS Crush <http://the-echoplex.net/csscrush/>`_. If you make references to CSS variables or other elements, the definition must be located within the same view output. E.g. A variable defined in ``elgg.css`` cannot be referenced in a separate CSS file like ``colorbox.css``.
+
```
in a future PR we will introduce a generic view with variables that gets injected into the csscrush api, this way you can use global (core) variables in all files..

jdalsem commented:
@mrclay what kind of test would you like to add? As this PR is otherwise ready to be merged

mrclay commented:
E.g. This removes vendor prefixes from our CSS so csscrush handles it. Would be nice to verify that happens.

hypeJunction commented:
I am now leaning towards scraping the theme and building anew with Bootstrap 4. We keep reinventing the wheel. If that happens, I don't see a reason to use CSSCrush yet, as B4 uses Sass.

jdalsem commented:
B4 'could' use SASS, but SASS and php do not work together so could. It would be cool to be able to preprocess css with a tool as a feature in Elgg. That way plugin devs can use css superglobals or mixins. The featurelist of CSS crush almost matches that of SASS, that is why i eventually suggested to use CSS Crush instead of SASS/LESS. There are some php-sass libraries, but those are not that complete. 

Theme builders can use sass for their own development of themes of course, but i would like to see preprocessing as part of Elgg core. The IMHO best solution for that is still CSS Crush

jdalsem committed:
feat(css): CSS is pre-processed by css-crush, all /cache output is filterable

All `/cache` output is now filtered by either the `simplecache:generate` or
`cache:generate` hook, depending on whether Simplecache is enabled.

css-crush is used to pre-process all CSS `/cache` output.

Fixes #10625
Refs #10316
